File: LeetCode/206.reverse-linked-list.cpp
/*
 * @lc app=leetcode id=206 lang=cpp
 *
 * [206] Reverse Linked List
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class SolutionA
{
public:
    ListNode *reverseList(ListNode *head)
    {
        if (head == nullptr || head->next == nullptr)
            return head;
        ListNode *left = reverseList(head->next);
        head->next->next = head;
        head->next = nullptr;
        return left;
    }
};

class SolutionB
{
public:
    ListNode *reverseList(ListNode *head)
    {
        ListNode dummy(0);
        while (head != nullptr)
        {
            ListNode *ele = head;
            head = head->next;
            ele->next = dummy.next;
            dummy.next = ele;
        }
        return dummy.next;
    }
};

class SolutionC
{
public:
    ListNode *reverseList(ListNode *head)
    {
        if (head == nullptr)
            return head;
        ListNode *cur = head->next;
        head->next = nullptr;
        while (cur != nullptr)
        {
            ListNode *ele = cur;
            cur = cur->next;
            ele->next = head;
            head = ele;
        }
        return head;
    }
};

File: LeetCode/java/206.reverse-linked-list.java
/*
 * @lc app=leetcode id=206 lang=java
 *
 * [206] Reverse Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode() {} ListNode(int val) { this.val = val; } ListNode(int val,
 * ListNode next) { this.val = val; this.next = next; } }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null)
            return head;
        ListNode res = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return res;
    }
}

class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode dummy = new ListNode(0);
        while (head != null) {
            ListNode tmp = head.next;
            head.next = dummy.next;
            dummy.next = head;
            head = tmp;
        }
        return dummy.next;
    }
}
// @lc code=end
